plugins/languages/treesitter/treesitter.nix

```diff
@@ -6,210 +6,235 @@
-        '')
-        + ''
-          require('nvim-treesitter.configs').setup(${helpers.toLuaObject tsOptions})
-        ''
-        + (optionalString (cfg.languageRegister != { }) ''
-          __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
-
-          for parser_name, ft in pairs(__parserFiletypeMappings) do
-            require('vim.treesitter.language').register(parser_name, ft)
-          end
-        '');
-
-      extraFiles = mkIf cfg.nixvimInjections {
-        "queries/nix/injections.scm" = ''
-          ;; extends
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @lua)
-              (indented_string_expression (string_fragment) @lua)
-            ]
-            (#match? @_path "^extraConfigLua(Pre|Post)?$"))
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @vim)
-              (indented_string_expression (string_fragment) @vim)
-            ]
-            (#match? @_path "^extraConfigVim(Pre|Post)?$"))
-        '';
-      };
+    nixGrammars = mkOption {
+      type = types.bool;
+      default = true;
+      description = "Whether to automatically install grammar packages.";
+    };
 
-      extraPlugins =
-        if cfg.nixGrammars then [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ] else [ cfg.package ];
-      extraPackages = with pkgs; [
-        tree-sitter
-        nodejs
-        cfg.gccPackage
-      ];
+    nixvimInjections = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''Whether to enable Nixvim injections, such as recognising extraConfigLua as lua.'';
+    };
+  };
 
-      opts = mkIf cfg.folding {
-        foldmethod = "expr";
-        foldexpr = "nvim_treesitter#foldexpr()";
-      };
+  callSetup = false;
+  extraConfig = cfg: {
+    extraConfigLua =
+      (optionalString (cfg.settings.parser_install_dir != null) ''
+        vim.opt.runtimepath:append("${cfg.settings.parser_install_dir}")
+      '')
+      + ''
+        require('nvim-treesitter.configs').setup(${helpers.toLuaObject cfg.settings})
+      ''
+      + (optionalString (cfg.languageRegister != { }) ''
+        __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
+
+        for parser_name, ft in pairs(__parserFiletypeMappings) do
+          require('vim.treesitter.language').register(parser_name, ft)
+        end
+      '');
+
+    extraFiles = mkIf cfg.nixvimInjections {
+      "queries/nix/injections.scm" = readFile ./queries/nix/injections.scm;
     };
+
+    extraPlugins = mkMerge [
+      (mkIf cfg.nixGrammars [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ])
+      (mkIf (!cfg.nixGrammars) [ cfg.package ])
+    ];
+
+    extraPackages = with pkgs; [
+      tree-sitter
+      nodejs
+      cfg.gccPackage
+    ];
```

Comment:
@GaetanLepage should this be upstreamed as nixpkgs dependencies?

I.e. `pkgs.tree-sitter` and `pkgs.nodejs`.

Reply:
Had a [discussion on matrix](https://matrix.to/#/!iBrKeqhFRZaBBGodLH:matrix.org/$YTqJrZwgfeGRswYV1W9TWDoCkVE4K-4vr09jAwU96dw?via=matrix.org&via=vhack.eu&via=gandom.nl)

- `pkgs.tree-sitter` shouldn't be needed as it is already bundled with neovim.
- `pkgs.nodejs` should be configurable by a nullable package option (like `cfg.gccPackage` is)
- `cfg.nodejsPackage` and `cfg.gccPackage` should probably default to null when `cfg.nixGrammars` is enabled? (This is a breaking change though, not 100% decided if we should do it) 

```nix
extraPackages = [
  cfg.nodejsPackage
  cfg.gccPackage
];
```

Reply:
If nodejs is just a requirement at runtime to build the grammar packages then I think it makes sense to follow in the same lines as the gccPackage.

Reply:
Yes, exactly

Reply:
>     * `pkgs.tree-sitter` shouldn't be needed as it is already bundled with neovim.

Seeing when we remove this package. Are we sure it's not needed for !nixGrammars?

```bash
- WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
```

Reply:
Based on that warning I'd say it is required. In that case I think we should handle it the same as nodejs and gcc.